Kernel/Devices/Loop/LoopDevice.cpp

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2024, Liav A. <liavalb@hotmail.co.il>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#include <AK/Singleton.h>
+#include <Kernel/Devices/DeviceManagement.h>
+#include <Kernel/Devices/Loop/LoopDevice.h>
+#include <Kernel/Sections.h>
+#include <Kernel/Tasks/WorkQueue.h>
+
+namespace Kernel {
+
+static Singleton<SpinlockProtected<LoopDevice::List, LockRank::None>> s_all_instances;
+static Singleton<SpinlockProtected<LoopDevice::ExposedList, LockRank::None>> s_all_exposed_instances;
+static Atomic<u64> s_loop_device_id { 0 };
+
+SpinlockProtected<LoopDevice::List, LockRank::None>& LoopDevice::all_instances()
+{
+    return s_all_instances;
+}
+
+SpinlockProtected<LoopDevice::ExposedList, LockRank::None>& LoopDevice::all_exposed_instances()
+{
+    return s_all_exposed_instances;
+}
+
+void LoopDevice::remove(Badge<DeviceControlDevice>)
+{
+    LoopDevice::all_exposed_instances().with([&](auto&) {
+        m_userspace_exposed_list_node.remove();
+    });
+}
+
+void LoopDevice::will_be_destroyed()
+{
+    m_backing_custody->inode().loop_device_attached({}).with([](auto& loop_device_attached) {
+        VERIFY(loop_device_attached);
+        loop_device_attached = false;
+    });
+    BlockDevice::will_be_destroyed();
+}
+
+bool LoopDevice::unref() const
+{
+    bool did_hit_zero = LoopDevice::all_instances().with([&](auto&) {
+        if (deref_base())
+            return false;
+        m_list_node.remove();
+        const_cast<LoopDevice&>(*this).revoke_weak_ptrs();
+        return true;
+    });
+    if (did_hit_zero) {
+        const_cast<LoopDevice&>(*this).will_be_destroyed();
+        delete this;
+    }
+    return did_hit_zero;
+}
+
+static ErrorOr<NonnullRefPtr<LoopDevice>> create_or_find_with_already_attached_inode(LoopDevice::ExposedList& locked_all_exposed_instances, LoopDevice::List& locked_all_instances_list, Custody& custody)
+{
+    // NOTE: Don't allow creating a new device if we already have such
+    // device with associated Inode, as it will make it possible to create
+    // multiple devices with the same Inode, which is bad for us.
+    // The main problem with this is that different FileSystem instances
+    // can be created that will be directed to the same Inode, therefore
+    // creating a possible filesystem corruption.
```

Comment:
I disagree with this logic - I think the user should be able to set up multiple loop devices based on the same inode, and they should be able to corrupt the backing store for the loop devices by mounting the same filesystem multiple times. This is just the effect of their actions, not something that happened inadvertently because of a bug in our code.

I don't feel we need to be this pedantic about potentially destructive actions by the user. They should think about what they're doing, and superusers should be able to do what they want to do :^)

For reference, Linux allows setting up multiple loop devices on the same inode as well.